### PR TITLE
Add 0-dim support for binary mkldnn ops

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52614 Avoid dispatch overhead in call to mkldnn convolution
* #52613 Include max pool in fusion groups
* #52513 Extend subgraph utils to cover merging a node following a subgraph
* #52512 Remove unused subgraph vmap api
* #51924 Only functionalize ops which we want to include in mkldnn group
* #51923 Add a filter to remove mutation
* #51922 Dont error on 0-dim in convolution
* **#51921 Add 0-dim support for binary mkldnn ops**
* #51736 Handle mkldnn broadcasting in mkldnn fuser
* #51600 Add MKLDNN fuser
* #52786 [NNC] Make NNC sanitize input names
* #51484 Add support for linear in mkldnn fusion
* #51483 Move conv to mkldnn

Differential Revision: [D26696696](https://our.internmc.facebook.com/intern/diff/D26696696)

### CODE REVIEW

File: aten/src/ATen/native/mkldnn/BinaryOps.cpp [hunk 1]
```diff
@@ -63,6 +93,10 @@ Tensor& mkldnn_add_out(
 }
 
 Tensor mkldnn_add(const Tensor& self, const Tensor& other, Scalar alpha) {
+  if (self.numel() == 0 | other.numel() == 0) {
```
Comment (Krovatkin) [inline aten/src/ATen/native/mkldnn/BinaryOps.cpp:96]: `||` ?

File: aten/src/ATen/native/mkldnn/BinaryOps.cpp [hunk 2]
```diff
@@ -102,6 +136,9 @@ Tensor& mkldnn_mul_out(Tensor& result, const Tensor& self, const Tensor& other)
 }
 
 Tensor mkldnn_mul(const Tensor& self, const Tensor& other) {
+  if (self.numel() == 0 | other.numel() == 0) {
```
Comment (Krovatkin) [inline aten/src/ATen/native/mkldnn/BinaryOps.cpp:139]: `||` ?

File: aten/src/ATen/native/mkldnn/BinaryOps.cpp [hunk 3]
```diff
@@ -46,6 +47,35 @@ Tensor& mkldnn_mul_(Tensor& self, const Tensor& other) {
 namespace at {
 namespace native {
 
+Tensor emptyBinaryOp(const Tensor& self, const Tensor& other) {
+  if (!self.requires_grad() && !other.requires_grad()) {
+    auto out_size = infer_size(self.sizes(), other.sizes());
```
Comment (Krovatkin) [inline aten/src/ATen/native/mkldnn/BinaryOps.cpp:52]: what if shapes are incompatible? Does `infer_size` throw?

File: aten/src/ATen/native/mkldnn/BinaryOps.cpp [hunk 4]
```diff
@@ -46,6 +47,35 @@ Tensor& mkldnn_mul_(Tensor& self, const Tensor& other) {
...
+    auto out_dtype = promoteTypes(
```
Comment (Krovatkin) [inline aten/src/ATen/native/mkldnn/BinaryOps.cpp:53]: is it always possible to promote?

File: aten/src/ATen/native/mkldnn/BinaryOps.cpp [hunk 5]
```diff
@@ -46,6 +47,35 @@ Tensor& mkldnn_mul_(Tensor& self, const Tensor& other) {
...
+        self.options().device_opt(),
+        self.options().pinned_memory_opt());
+  } else {
+    TORCH_CHECK(
+        false,
+        "MKLDNN does not support Binary Ops with a 0-dimension Tensor in training");
+  }
+}
+
+Tensor empty_strided_from_mkldnn(const Tensor& mkldnn_tensor) {
```
Comment (Krovatkin) [inline aten/src/ATen/native/mkldnn/BinaryOps.cpp:72]: where is this used?

File: test/test_mkldnn.py [hunk 6]
```diff
@@ -514,6 +514,28 @@ def test_mul(self):
...
+        out_relu = torch.relu(y)
+        out_relu_mkldnn = torch.relu(y.to_mkldnn()).to_dense()
+        self.assertEqual(out_relu, out_relu_mkldnn)
+
+        out_mul = x * y
+        out_mul_mkldnn = (x.to_mkldnn() * y.to_mkldnn()).to_dense()
+        self.assertEqual(out_mul, out_mul_mkldnn)
+
+        out_add = x + y
+        out_add_mkldnn = (x.to_mkldnn() + y.to_mkldnn()).to_dense()
```
Comment (Krovatkin) [inline test/test_mkldnn.py:531]: might not be in the scope of this PR but what happens in case of conv and linear?

**Review by Krovatkin — COMMENTED** [review]
some questions

File: aten/src/ATen/native/mkldnn/BinaryOps.cpp (hunk 3 above)
Comment (eellison) [reply aten/src/ATen/native/mkldnn/BinaryOps.cpp:52]: yea, i'm 99% sure I checked and it does. i'll add a test

File: aten/src/ATen/native/mkldnn/BinaryOps.cpp (hunk 4 above)
Comment (eellison) [reply aten/src/ATen/native/mkldnn/BinaryOps.cpp:53]: yes. promoteTypes always succeeds. MKLDNN only supports bfloat16 & float as well.

File: test/test_mkldnn.py (hunk 6 above)
Comment (eellison) [reply test/test_mkldnn.py:531]: this is in follow up PRs.

File: aten/src/ATen/native/mkldnn/BinaryOps.cpp (hunk 3 above)
Comment (eellison) [reply aten/src/ATen/native/mkldnn/BinaryOps.cpp:52]: this test was added

**Review by Krovatkin — APPROVED** [review]
:shipit:
